### PR TITLE
ONT-566 add admin role;fix commitDpos bug;update blacknode logic;change view to uint32;fix getOngBalance

### PR DESCRIPTION
**Pull request by siovanus**

### CODE REVIEW

File: smartcontract/service/native/governance/param.go [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+ * along with The ontology.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package governance
+
+import (
+	"github.com/ontio/ontology/common"
+	"github.com/ontio/ontology/common/serialization"
+	"github.com/ontio/ontology/errors"
+	"io"
```
Comment (Honglei-Cong) [inline smartcontract/service/native/governance/param.go]: format import

File: smartcontract/service/native/governance/utils.go [hunk 2]
```diff
@@ -232,9 +235,53 @@ func InitConfigBytes() []byte {
...
+func splitCurve(pos uint64, avg uint64, yita uint32) uint64 {
+	xi := PRECISE * uint64(yita) * 2 * pos / (avg * 10)
 	index := xi / (PRECISE / 10)
 	s := ((Yi[index+1]-Yi[index])*xi + Yi[index]*Xi[index+1] - Yi[index+1]*Xi[index]) / (Xi[index+1] - Xi[index])
 	return s
 }
+
+func GetUint32Bytes(view uint32) []byte {
+	var p [4]byte
+	binary.LittleEndian.PutUint32(p[:], view)
```
Comment (Honglei-Cong) [inline smartcontract/service/native/governance/utils.go]: please use serialization interface.  

same comments for the following GetBytesUint32.

File: smartcontract/service/native/governance/utils.go [hunk 3]
```diff
@@ -26,7 +26,9 @@ import (
 	"math"
 	"math/big"
 
+	"encoding/binary"
```
Comment (Honglei-Cong) [inline smartcontract/service/native/governance/utils.go]: don't use binary, please change to serialization.

File: smartcontract/service/native/governance/param.go [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+	}
+	return nil
+}
+
+func (this *RegisterSyncNodeParam) Deserialize(r io.Reader) error {
+	peerPubkey, err := serialization.ReadString(r)
+	if err != nil {
+		return errors.NewDetailErr(err, errors.ErrNoCode, "serialization.ReadString, deserialize peerPubkey error!")
+	}
+	this.PeerPubkey = peerPubkey
```
Comment (Honglei-Cong) [inline smartcontract/service/native/governance/param.go]: update this.[*] after all data members are deserialized successfully.

File: smartcontract/service/native/governance/param.go [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+	Address        common.Address
+	PeerPubkeyList []string
+	PosList        []uint32
+}
+
+func (this *VoteForPeerParam) Serialize(w io.Writer) error {
+	if err := this.Address.Serialize(w); err != nil {
+		return errors.NewDetailErr(err, errors.ErrNoCode, "serialization.WriteString, serialize address error!")
+	}
+	if err := serialization.WriteVarUint(w, uint64(len(this.PeerPubkeyList))); err != nil {
```
Comment (Honglei-Cong) [inline smartcontract/service/native/governance/param.go]: serialization.WriteUint32 ? 
or serialization.WriteUint64 ?

File: smartcontract/service/native/governance/param.go [hunk 6]
```diff
@@ -0,0 +1,554 @@
...
+		return errors.NewDetailErr(err, errors.ErrNoCode, "serialization.WriteVarUint, serialize peerPubkeyList length error!")
+	}
+	for _, v := range this.PeerPubkeyList {
+		if err := serialization.WriteString(w, v); err != nil {
+			return errors.NewDetailErr(err, errors.ErrNoCode, "serialization.WriteString, serialize peerPubkey error!")
+		}
+	}
+	if err := serialization.WriteVarUint(w, uint64(len(this.PosList))); err != nil {
```
Comment (Honglei-Cong) [inline smartcontract/service/native/governance/param.go]: WriteUint64 ?

**Review by Honglei-Cong — CHANGES_REQUESTED** [review]

File: smartcontract/service/native/governance/param.go [hunk 7]
```diff
@@ -0,0 +1,554 @@
...
+	n, err := serialization.ReadVarUint(r, 0)
+	if err != nil {
+		return errors.NewDetailErr(err, errors.ErrNoCode, "serialization.ReadVarUint, deserialize peerPubkeyList length error!")
+	}
+	for i := 0; uint64(i) < n; i++ {
+		k, err := serialization.ReadString(r)
+		if err != nil {
+			return errors.NewDetailErr(err, errors.ErrNoCode, "serialization.ReadString, deserialize peerPubkey error!")
+		}
+		this.PeerPubkeyList = append(this.PeerPubkeyList, k)
```
Comment (Honglei-Cong) [inline smartcontract/service/native/governance/param.go]: make an empty array firstly,
then append to the empty array.

File: smartcontract/service/native/governance/param.go [hunk 8]
```diff
@@ -0,0 +1,554 @@
...
+	m, err := serialization.ReadVarUint(r, 0)
+	if err != nil {
+		return errors.NewDetailErr(err, errors.ErrNoCode, "serialization.ReadVarUint, deserialize posList length error!")
+	}
+	for i := 0; uint64(i) < m; i++ {
+		k, err := serialization.ReadUint32(r)
+		if err != nil {
+			return errors.NewDetailErr(err, errors.ErrNoCode, "serialization.ReadString, deserialize pos error!")
+		}
+		this.PosList = append(this.PosList, k)
```
Comment (Honglei-Cong) [inline smartcontract/service/native/governance/param.go]: same comment as the above one.

**Review by Honglei-Cong — CHANGES_REQUESTED** [review]

File: smartcontract/service/native/governance/param.go [hunk 9]
```diff
@@ -0,0 +1,554 @@
...
+	}
+	for i := 0; uint64(i) < n; i++ {
+		k, err := serialization.ReadString(r)
+		if err != nil {
+			return errors.NewDetailErr(err, errors.ErrNoCode, "serialization.ReadString, deserialize peerPubkey error!")
+		}
+		this.PeerPubkeyList = append(this.PeerPubkeyList, k)
+	}
+
+	m, err := serialization.ReadVarUint(r, 0)
```
Comment (Honglei-Cong) [inline smartcontract/service/native/governance/param.go]: check m == n ?

File: smartcontract/service/native/governance/param.go [hunk 10]
```diff
@@ -0,0 +1,554 @@
...
+	}
+	if err := serialization.WriteVarUint(w, uint64(len(this.PeerPubkeyList))); err != nil {
+		return errors.NewDetailErr(err, errors.ErrNoCode, "serialization.WriteVarUint, serialize peerPubkeyList length error!")
+	}
+	for _, v := range this.PeerPubkeyList {
+		if err := serialization.WriteString(w, v); err != nil {
+			return errors.NewDetailErr(err, errors.ErrNoCode, "serialization.WriteString, serialize peerPubkey error!")
+		}
+	}
+	if err := serialization.WriteVarUint(w, uint64(len(this.WithdrawList))); err != nil {
```
Comment (Honglei-Cong) [inline smartcontract/service/native/governance/param.go]: similar comment as VoteForPeer

File: smartcontract/service/native/governance/param.go [hunk 11]
```diff
@@ -0,0 +1,554 @@
...
+	}
+	return nil
+}
+
+func (this *Configuration) Deserialize(r io.Reader) error {
+	n, err := serialization.ReadUint32(r)
+	if err != nil {
+		return errors.NewDetailErr(err, errors.ErrNoCode, "serialization.ReadUint32, deserialize n error!")
+	}
+	this.N = n
```
Comment (Honglei-Cong) [inline smartcontract/service/native/governance/param.go]: same comment as above, 
update this after all data members deserialized successfully.

**Review by Honglei-Cong — CHANGES_REQUESTED** [review]

File: smartcontract/service/native/governance/governance.go [hunk 12]
```diff
@@ -149,15 +133,46 @@ func InitConfig(native *native.NativeService) ([]byte, error) {
 		return utils.BYTE_FALSE, errors.NewErr("initConfig. initConfig is already executed!")
 	}
 
+	//init globalParam
```
Comment (laizy) [inline smartcontract/service/native/governance/governance.go:138]: all the config value got from external environment should be serialized in argument.

File: smartcontract/service/native/init/init.go [hunk 13]
```diff
@@ -55,3 +57,14 @@ func InitBytes(addr common.Address, method string) []byte {
 	init.Serialize(bf)
 	return bf.Bytes()
 }
+
+func InitConfigBytes(addr common.Address, method string) []byte {
+	configbf := new(bytes.Buffer)
+	if err := config.DefConfig.Genesis.VBFT.Serialize(configbf); err != nil {
+		fmt.Println("vbft config serialize failed")
```
Comment (laizy) [inline smartcontract/service/native/init/init.go:64]: use log

File: smartcontract/service/native/governance/governance.go [hunk 14]
```diff
@@ -912,26 +1175,37 @@ func CommitDpos(native *native.NativeService) ([]byte, error) {
...
+	err = executeCommitDpos(native, contract, config)
+	if err != nil {
+		return utils.BYTE_FALSE, errors.NewDetailErr(err, errors.ErrNoCode, "executeCommitDpos, executeCommitDpos error!")
+	}
 
 	return utils.BYTE_TRUE, nil
 }
 
 func executeCommitDpos(native *native.NativeService, contract common.Address, config *Configuration) error {
+	fmt.Println("############################")
```
Comment (laizy) [inline smartcontract/service/native/governance/governance.go]: remove this

**Review by laizy — CHANGES_REQUESTED** [review]

**Review by laizy — APPROVED** [review]

**Pull request merged**